### PR TITLE
Update setting lookup from environment variables

### PR DESCRIPTION
**Pull request by fmoor**
fix https://github.com/edgedb/edgedb-go/issues/178

### CODE REVIEW

**Comment by fmoor** [comment]
This PR is blocking on https://github.com/edgedb/shared-client-testcases/pull/2

File: credentials.go [hunk 1]
```diff
@@ -25,13 +25,13 @@ import (
...
-	password       OptionalStr
-	certData       OptionalBytes
-	verifyHostname OptionalBool
+	host        OptionalStr
+	port        OptionalInt32
+	user        string
+	database    OptionalStr
+	password    OptionalStr
+	certData    OptionalBytes
+	tlsSecurity OptionalStr
```
[the remark this answers is not in the document]
Comment (fmoor) [reply credentials.go:34]: There is no such thing as an enum in go. We could subtype string if you thing that is worth it but currently the go client maps enums to string.

[the remark this answers is not in the document]
Comment (tailhook) [reply credentials.go:34]: I mean something like this: https://www.educative.io/edpresso/what-is-an-enum-in-golang
But yeah, it doesn't add to correctness in go, and is not used in performance-sensitive code, so strings are fine.

File: connutils.go [hunk 2]
```diff
@@ -45,8 +45,59 @@ type connConfig struct {
...
+		}
+	}
+
+	tlsConfig := &tls.Config{
+		RootCAs:    roots,
+		NextProtos: []string{"edgedb-binary"},
+	}
+
+	security, err := getEnvVarSetting(
+		"EDGEDB_CLIENT_SECURITY", "default", "default", "insecure_dev_mode")
```
Comment (elprans) [inline connutils.go:74]: Add `strict`, which is the default.

**Review by elprans — APPROVED** [review]
LGTM, but please add the `strict` option to the tls security enum.

File: connutils.go (hunk 2 above)
Comment (fmoor) [reply connutils.go:74]: I'm confused. This is the replacement for `EDGEDB_INSECURE_DEV_MODE`. Is the default really `strict` TLS?

Comment (fmoor) [reply connutils.go:74]: Shouldn't the default be to have no effect? Then whatever `tls_security` workouts out to be (who's default is `strict`) will be the end result.

Comment (elprans) [reply connutils.go:74]: Hm, you're right. The default is the same as `EDGEDB_CLIENT_TLS_SECURITY`, but `strict` should still be an option.

File: connutils.go [hunk 3]
```diff
@@ -45,8 +45,60 @@ type connConfig struct {
...
+	}
+
+	tlsConfig := &tls.Config{
+		RootCAs:    roots,
+		NextProtos: []string{"edgedb-binary"},
+	}
+
+	security, err := getEnvVarSetting(
+		"EDGEDB_CLIENT_SECURITY", "default", "default", "insecure_dev_mode",
+		"strict")
```
Comment (elprans) [inline connutils.go:75]: `strict` should enable `tls_security=strict` (or error if it's set to a conflicting value).

File: connutils.go [hunk 4]
```diff
@@ -615,11 +608,34 @@ func (r *configResolver) config(opts *Options) (*connConfig, error) {
...
+		"default", "insecure_dev_mode", "strict")
+	if err != nil {
+		return nil, err
+	}
+
+	switch security {
+	case "default":
+	case "insecure_dev_mode":
+		if tlsSecurity == "default" {
+			tlsSecurity = "insecure"
```
Comment (elprans) [inline connutils.go:621]: No-no, "default" means "default" for tlsSecurity also.

Comment (fmoor) [reply connutils.go:621]: I don't know how to interpret this. Can you write out something like a truth table for the interaction between these two values?

Comment (elprans) [reply connutils.go:621]:
```
CLIENT_SECURITY            CLIENT_TLS_SECURITY    result TLS_SECURITY
-------------------------  ---------------------  --------------------
default                    default                default
default                    insecure               insecure
default                    no_host_verification   no_host_verification
default                    strict                 strict
insecure_dev_mode          default                insecure
insecure_dev_mode          insecure               insecure
insecure_dev_mode          no_host_verification   no_host_verification
insecure_dev_mode          strict                 strict
strict                     default                strict
strict                     insecure               ERROR
strict                     no_host_verification   ERROR
strict                     strict                 strict
```

Comment (fmoor) [reply connutils.go:621]: I must be missing something. Where does the implementation differ from the table?

Comment (elprans) [reply connutils.go:621]: "default" `CLIENT_SECURITY` leads to "insecure" TLS mode instead of "default"

Comment (fmoor) [reply connutils.go:621]: The `CLIENT_SECURITY=default` case (line `618`) has an empty body which is a noop. Are you reading it as if that case falls through to the next case? switch cases don't fall through unless there is an explicit `fallthrough` statement.

Comment (elprans) [reply connutils.go:621]: Oh, yeah that was it, I was reading it as C. :+1:

**Pull request merged**